scormxblock/static/js/src/scormxblock.js

```diff
@@ -4,105 +4,165 @@ function ScormXBlock(runtime, element, settings) {
+    '404': 'Element is write only',
+    '405': 'Incorrect Data Type',
+  };
+
+  var values = [];
+  var LMSValues = {};
+
   var GetValue = function (cmi_element) {
-    var handlerUrl = runtime.handlerUrl(element, 'scorm_get_value');
 
-    var response = $.ajax({
-      type: "POST",
-      url: handlerUrl,
-      data: JSON.stringify({'name': cmi_element}),
-      async: false
-    });
-    response = JSON.parse(response.responseText);
-    console.log("Getvalue for " + cmi_element + " = " + response.value);
-    return response.value
+    if (settings.valid_read_data.includes(cmi_element)){
+      var handlerUrl = runtime.handlerUrl(element, 'scorm_get_values');
+
+      if (values.length===0) {
+        var response = $.ajax({
+          type: 'GET',
+          url: handlerUrl,
+          async: false,
+        });
+        values = JSON.parse(response.responseText);
+      }
+
+      var value = values[cmi_element];
+
+      if (value==undefined){
+        value = '';
+      }
+
+      console.log('Getvalue for ' + cmi_element + ' = ' + value);
+      return value
+    } else if (settings.valid_write_data.includes(cmi_element)) {
+      lastError = 404;
+    } else if (cmi_element.endsWith('_count')){
+      lastError = 203;
+    } else if (cmi_element.endsWith('_children')){
+      lastError = 202;
+    } else {
+      lastError = 401;
+    }
+    return '';
+
   };
 
   var SetValue = function (cmi_element, value) {
-    console.log("LMSSetValue " + cmi_element + " = " + value);
-    var handlerUrl = runtime.handlerUrl( element, 'scorm_set_value');
+
+    if (value == null || value == 'undefined'){
+      lastError = 405;
+    } else if (settings.valid_write_data.includes(cmi_element)) {
+      console.log('LMSSetValue ' + cmi_element + ' = ' + value);
+      LMSValues[cmi_element] = value
+      return 'true';
+    } else if (cmi_element.endsWith('_count') || cmi_element.endsWith('_children')){
+      lastError = 402;
+    } else if (settings.valid_read_data.includes(cmi_element)){
+      lastError = 403;
+    } else {
+      lastError = 401;
+    }
+    return 'false';
+
+  };
+
+  var StoreValues = function () {
+    var handlerUrl = runtime.handlerUrl( element, 'scorm_set_values');
 
     $.ajax({
-      type: "POST",
+      type: 'POST',
       url: handlerUrl,
-      data: JSON.stringify({'name': cmi_element, 'value': value}),
+      data: JSON.stringify(LMSValues),
       async: false,
       success: function(response){
-        if (typeof response.lesson_score != "undefined"){
-          $(".lesson_score", element).html(response.lesson_score);
+        if (typeof response.lesson_score != 'undefined'){
+          $('.lesson_score', element).html(response.lesson_score);
         }
-        $(".completion_status", element).html(response.completion_status);
+        $('.completion_status', element).html(response.completion_status);
+        LMSValues = {};
       }
     });
 
-    return "true";
+    return 'true';
```

Comment:
Does this need to be a string or could it be a boolean? @andrey-canon

Reply:
This doesn't make a difference in this method but I'm going to keep it for consistence